system/lib/pthread/cxa_thread_atexit.c

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2019 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ * Inspired by libcxxabi/src/cxa_thread_atexit.cpp.
+ * The main reasons we don't use that version direclty are:
+ * 1. We want to be able to use __cxa_thread_atexit in pure C programs
+ *    where libcxxabi is not linked in at all.
+ * 2. The libcxxabi relies on TLS variables and we use to free our TLS data
+ *    block on thread exit.  This would cause a chicken and egg issue
+ *    where the TLS variables in libcxxabi/src/cxa_thread_atexit.cpp would
+ *    be freed while porcessing the list of __cxa_thread_atexit handlers
+ */
+#include <assert.h>
+#include <stdbool.h>
+#include <stdatomic.h>
+#include <pthread.h>
+#include <stdlib.h>
+
+#include "libc.h"
+
+typedef void(*Dtor)(void*);
+
+typedef struct DtorList {
+  Dtor dtor;
+  void* obj;
+  struct DtorList* next;
+} DtorList;
+
+void run_dtors(void* arg) {
+  DtorList* dtors = (DtorList*)arg;
+  DtorList* head;
+  while ((head = dtors)) {
+    dtors = head->next;
+    head->dtor(head->obj);
+    free(head);
+  }
+}
+
+static pthread_key_t key;
+atomic_bool key_created = false;
```

Comment:
This should probably be a `thread_local` instead of an atomic bool.
```suggestion
static thread_local bool key_created = false;
```

Reply:
(this seems to fix `posixtest.test_pthread_exit_3_2` locally)

Reply:
IIUC pthread_key_create should only be called once from a single thread.. so this guard needs to be shared between thread, no?   

Also see the comment above for why we specifically can't use TLS variables here.

Reply:
I'm not sure if it should be shared across threads or if it should only be initialized once for a single thread. Given that the `dtors_alive` variable in `libcxxabi` was also thread local, I thought the latter approach was correct here (but I might be wrong).

> Also see the comment above for why we specifically can't use TLS variables here.

Ah yes, I read that comment to fast. Could the C extension `__thread` help here? AFAIK, that doesn't trigger the `run_dtors` destructor in `libcxxabi` and would correspond to the `dtors_alive` variable.
https://github.com/emscripten-core/emscripten/blob/4d864df0a57024d667e8db171aab2c3385d04c30/system/lib/libcxxabi/src/cxa_thread_atexit.cpp#L71-L72

Reply:
Basically we are using pthread_key_create *instead* of the TLS variables used by `emscripten/system/lib/libcxxabi/src/cxa_thread_atexit.cpp`.

There are essentially two ways to achieve TLS:   You can either use C/C++ __thread builtin, or you can do it at runtime with `pthread_key_create`.   We are using the `pthread_key_create` method here because we cannot use any or C++ TLS storage here because the space needed by the TLS variables themselves is freed using a `__cxa_thead_atexit` handler.. even if the TLS is just plain old data.

When one uses the `pthread_key_create` one needs to make sure that `pthread_key_create` is only run once.  Normally this is done using `pthread_once`, but that also can't be used in this case because it uses `pthread_cleanup_push`/`pop` which itself uses `__cxa_thead_atexit`.  I've tried to explain more in the commands above.  

For an example of how `pthread_once` is normally used to call `pthread_key_create` see: https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_key_create.html